### PR TITLE
Event formatting in month view refactored to introduce EventFormatter

### PR DESCRIPTION
**Pull request by 0xym**
Long term goal of my changes is to allow multi-line calendar events in MonthView (this should be configurable)
This commit decouples dealing with some formatting issues - like whether to show event time, or "more" indicator, from actual event drawing.

### CODE REVIEW

**Comment by 0xym** [comment]
I suppose it should be ready for testing now.
Do you have some automatic test suite? I haven't run any unit tests (nor found any). I have tested some obvious scenarios and it seems to me that everything works the same way. 
I also don't know what is your merging strategy - do you want to merge my chances only after there will be some functional changes?

**Comment by jspricke** [comment]
Thx for your pull request. You are changing quiet a bit of code, so I will probably need some more time to review it. I played around with it a little andfound one bug, can you fix that maybe?
etar

**Comment by 0xym** [comment]
Interesting. Unfortunately I won't have time to look into it until Tuesday. I hope it will be easy to reproduce it.

**Comment by 0xym** [comment]
I have started the work on multiline events on Friday. I've had initial code ready yesterday. Today I have fixed remaining bugs. It turns out that the bug you have mentioned should be fixed with this commit.
To turn off multiline events it is enough to set mMaxLinesInEvent to 1 (I have set it to 7 to so it would be almost impossible that it would be a real limit - but I didn't want to set it to a higher value, because increasing it doesn't come for free).

**Comment by 0xym** [comment]
the issues should be gone now

**Comment by jspricke** [comment]
I found an other instance of showing less events then available space. I have two full day events with a short title and three non full day. The first non full day has a rather long title (>47 chars in my case). The is that the two full day events and a `+3` are shown. If I cut the title at 46 chars, the beginning of it is shown as well (and a '+2'). I didn't went through the code yet, but maybe you have an idea why this happens?

Really nice work otherwise!

**Comment by 0xym** [comment]
I have looked briefly into the code and I think that t there is "off by one" error in line 1176. In reduceHeightOfEventsToOne the loop should be inclusive (condition should be "<=" )

**Comment by jspricke** [comment]
@0xym the "<=" solved it, thx! I pushed a fixup.

**Comment by 0xym** [comment]
I suppose that this part of changes is complete now (unless there are some bugs). There is one thing that might need to be added - some configuration. But I don't know what should be configurable, to what extent and where to put the configuration.

**Comment by 0xym** [comment]
For dealing with text formatting I'm using android.text.DynamicLayout. Text breaking can be improved, but only in API28 (Android 9) - there is DynamicLayout.Builder which can produce DynamicLayouts with selected breaking strategies (there are 3 to choose from now). Currently less than 1% of Android devices runs Pie, so I'm afraid that going this way won't make a big difference

File: src/com/android/calendar/month/MonthWeekEventsView.java [hunk 1]
```diff
@@ -103,9 +106,12 @@
     private static int EVENT_BOTTOM_PADDING = 1;
     private static int TODAY_HIGHLIGHT_WIDTH = 2;
     private static int SPACING_WEEK_NUMBER = 0;
+    private static int BORDER_SPACE;
```
Comment (jspricke) [inline src/com/android/calendar/month/MonthWeekEventsView.java]: I think these should be lower case (as they are not constants), see: https://source.android.com/setup/contribute/code-style#follow-field-naming-conventions

(also it's not consistent in the file).

File: src/com/android/calendar/month/MonthWeekEventsView.java [hunk 2]
```diff
@@ -785,152 +753,912 @@ protected int addChipOutline(FloatRef lines, int count, int x, int y) {
         return count;
     }
 
+    protected class DayEventSorter {
+        private final EventFormat virtualFormat = new EventFormat(0,0);
```
Comment (jspricke) [inline src/com/android/calendar/month/MonthWeekEventsView.java]: add a space after the comma.

File: src/com/android/calendar/month/MonthWeekEventsView.java [hunk 3]
```diff
@@ -785,152 +753,912 @@ protected int addChipOutline(FloatRef lines, int count, int x, int y) {
...
+                        break;
+                    }
+                    index = getNextIndex(indexedEvents, index);
+                }
+            }
+            addVirtualEvents(indexedEvents, index);
+        }
+
+        protected void addVirtualEvents(FormattedEventBase[] indexedEvents, int initialIndex)  {
+            for (int index = initialIndex; index < mMinItems; ++index) {
```
Comment (jspricke) [inline src/com/android/calendar/month/MonthWeekEventsView.java]: I would go for `index++` (seems to be more common in the code base).

File: src/com/android/calendar/month/MonthWeekEventsView.java [hunk 4]
```diff
@@ -785,152 +753,912 @@ protected int addChipOutline(FloatRef lines, int count, int x, int y) {
...
-         *
-         * e. "+ more" line (mExtrasHeight)
-         *
-         * f. EVENT_BOTTOM_PADDING (overlaps EVENT_LINE_PADDING)
-         *   (bottom of box)
+    protected class DayEventFormatter {
+        private ArrayList<FormattedEventBase> mEventDay;
+        private boolean mShowTimes;
+        private int mDay;
+        //info
```
Comment (jspricke) [inline src/com/android/calendar/month/MonthWeekEventsView.java]: What should the comment mean?

File: src/com/android/calendar/month/MonthWeekEventsView.java (hunk 1 above)
Comment (0xym) [reply src/com/android/calendar/month/MonthWeekEventsView.java]: fixed

File: src/com/android/calendar/month/MonthWeekEventsView.java (hunk 2 above)
Comment (0xym) [reply src/com/android/calendar/month/MonthWeekEventsView.java]: done

File: src/com/android/calendar/month/MonthWeekEventsView.java (hunk 3 above)
Comment (0xym) [reply src/com/android/calendar/month/MonthWeekEventsView.java]: using pre incrementation is more efficient (but compiler should figure it out and change post-incrementation to pre-incrementation whenever it has no side effect)
I have changed the code to follow your suggestion

File: src/com/android/calendar/month/MonthWeekEventsView.java (hunk 4 above)
Comment (0xym) [reply src/com/android/calendar/month/MonthWeekEventsView.java]: comment changed to a more meaningful one

**Pull request merged**